### PR TITLE
Further improve D2k spiceblooms

### PR DESCRIPTION
**Pull request by reaperrr**
This implements the last missing aspects of spiceblooms, namely only growing/'spawning' on spice-free sand (unless pre-placed in map editor) and displaying the animation of blowing spice into the air when spawned.

Additionally, spice particles now properly target cells free of spice when possible.

Closes https://github.com/OpenRA/OpenRA/issues/9866.

### CODE REVIEW

**Comment by reaperrr** [comment]
I intentionally kept this out from my previous PR because this leaves the 'fix' area and goes more into 'feature' territory and might need some polish because it's a little hacky.

Additionally, `var cells` in `SpiceBloom` probably still returns `null` all the time, but I don't know how to fix that.

**Comment by pchote** [comment]
Do you really even want that random walk?  That behaviour is used in the other mods because it nicely simulates random growth from a central tree... in this case you're throwing the resources through the air, so you only really care that it will land in an empty cell within some reasonable circular range.

**Comment by reaperrr** [comment]
Updated.

**Comment by Mailaender** [comment]
Very nice. Great attention to detail. This scales far better than infinite growth. :+1:

File: OpenRA.Mods.D2k/Traits/SpiceBloom.cs [hunk 1]
```diff
@@ -74,10 +81,18 @@ public SpiceBloom(ActorInitializer init, SpiceBloomInfo info)
 			respawnTicks = self.World.SharedRandom.Next(info.RespawnDelay[0], info.RespawnDelay[1]);
 			growTicks = self.World.SharedRandom.Next(info.GrowthDelay[0], info.GrowthDelay[1]);
 			anim.Animation.Play(info.GrowthSequences[0]);
+
+			spawnCell = self.World.Map.CellContaining(self.CenterPosition);
 		}
 
 		public void Tick(Actor self)
 		{
+			if (!self.World.Map.Contains(spawnCell))
```
Comment (pchote) [inline OpenRA.Mods.D2k/Traits/SpiceBloom.cs]: It would be better to use `self.Location` instead of caching `self.World.Map.CellContaining(self.CenterPosition)`.  The latter will break if the actor moves for whatever reason.

File: OpenRA.Mods.D2k/Traits/SpiceBloom.cs [hunk 2]
```diff
@@ -97,10 +112,10 @@ public void Killed(Actor self, AttackInfo e)
 
 			for (var i = 0; pieces > i; i++)
 			{
-				var cells = OpenRA.Traits.Util.RandomWalk(self.Location, self.World.SharedRandom);
-				var cell = cells.Take(info.Range).SkipWhile(p => resLayer.GetResource(p) == resType && resLayer.IsFull(p)).Cast<CPos?>().RandomOrDefault(self.World.SharedRandom);
+				var cells = self.World.Map.FindTilesInAnnulus(self.World.Map.CellContaining(self.CenterPosition), 1, info.Range);
```
Comment (pchote) [inline OpenRA.Mods.D2k/Traits/SpiceBloom.cs]: Why change from `self.Location`?

**Comment by pchote** [comment]
We'll want this for prep, right?

**Comment by penev92** [comment]
I guess so.

File: OpenRA.Mods.D2k/Traits/SpiceBloom.cs [hunk 3]
```diff
@@ -34,6 +37,9 @@ public class SpiceBloomInfo : ITraitInfo, Requires<RenderSpritesInfo>
 
 		public readonly string ResourceType = "Spice";
 
+		[Desc("Spice blooms only grow on this terrain type.")]
+		public readonly string GrowthTerrainType = "Sand";
```
Comment (penev92) [inline OpenRA.Mods.D2k/Traits/SpiceBloom.cs]: Since we have 1 actor that uses the default and 1 that removes the default, I'd rather we default to "" (or `null`, whichever is appropriate according to our standards) and put "Sand" on the actor that needs it.

File: mods/d2k/weapons.yaml [hunk 4]
```diff
@@ -820,3 +820,12 @@ BloomExplosion:
 		DamageTypes: Prone50Percent, TriggerProne, ExplosionDeath
 		AffectsParent: true
 
+BloomSpawn:
+	Range: 0c1
+	Projectile: Bullet
+		Speed: 1c0
+		Blockable: false
+		Image: null
+	Warhead@3Eff: CreateEffect
```
Comment (penev92) [inline mods/d2k/weapons.yaml]: This isn't the 3rd warhead on this weapon though.

**Comment by penev92** [comment]
Looks good ingame.
:+1:  with the two nits above.

File: OpenRA.Mods.D2k/Traits/SpiceBloom.cs (hunk 1 above)
Comment (reaperrr) [inline OpenRA.Mods.D2k/Traits/SpiceBloom.cs]: Done.

File: OpenRA.Mods.D2k/Traits/SpiceBloom.cs (hunk 2 above)
Comment (reaperrr) [inline OpenRA.Mods.D2k/Traits/SpiceBloom.cs]: Probably just misread it as self.CenterPosition originally. Anyway, fixed.

File: OpenRA.Mods.D2k/Traits/SpiceBloom.cs (hunk 3 above)
Comment (reaperrr) [inline OpenRA.Mods.D2k/Traits/SpiceBloom.cs]: Done.

File: mods/d2k/weapons.yaml (hunk 4 above)
Comment (reaperrr) [inline mods/d2k/weapons.yaml]: Fixed.

**Comment by reaperrr** [comment]
Updated.

File: OpenRA.Mods.D2k/Traits/SpiceBloom.cs [hunk 5]
```diff
@@ -34,6 +37,9 @@ public class SpiceBloomInfo : ITraitInfo, Requires<RenderSpritesInfo>
 
 		public readonly string ResourceType = "Spice";
 
+		[Desc("Spice blooms only grow on this terrain type.")]
+		public readonly string GrowthTerrainType = null;
```
Comment (pchote) [inline OpenRA.Mods.D2k/Traits/SpiceBloom.cs]: Can we make this an array instead, for flexibility?

File: OpenRA.Mods.D2k/Traits/SpiceBloom.cs [hunk 6]
```diff
@@ -97,10 +109,10 @@ public void Killed(Actor self, AttackInfo e)
 
 			for (var i = 0; pieces > i; i++)
 			{
-				var cells = OpenRA.Traits.Util.RandomWalk(self.Location, self.World.SharedRandom);
-				var cell = cells.Take(info.Range).SkipWhile(p => resLayer.GetResource(p) == resType && resLayer.IsFull(p)).Cast<CPos?>().RandomOrDefault(self.World.SharedRandom);
+				var cells = self.World.Map.FindTilesInAnnulus(self.Location, 1, info.Range);
```
Comment (pchote) [inline OpenRA.Mods.D2k/Traits/SpiceBloom.cs]: You can pull this and the `SkipWhile` outside the loop to save some perf.

File: OpenRA.Mods.D2k/Traits/SpiceBloom.cs [hunk 7]
```diff
@@ -34,6 +37,9 @@ public class SpiceBloomInfo : ITraitInfo, Requires<RenderSpritesInfo>
 
 		public readonly string ResourceType = "Spice";
 
+		[Desc("Spice blooms only grow on this terrain type.")]
+		public readonly string GrowthTerrainType = null;
+
 		[Desc("The weapon to use for spice creation.")]
 		[WeaponReference]
 		public readonly string Weapon = "SpiceExplosion";
```
Comment (pchote) [inline OpenRA.Mods.D2k/Traits/SpiceBloom.cs]: Can we default this to null so that you don't need to override it with an empty entry in the spawnpoint?

File: OpenRA.Mods.D2k/Traits/SpiceBloom.cs [hunk 8]
```diff
@@ -78,6 +84,12 @@ public SpiceBloom(ActorInitializer init, SpiceBloomInfo info)
 
 		public void Tick(Actor self)
 		{
+			if (!self.World.Map.Contains(self.Location))
+				return;
+
+			if (info.GrowthTerrainTypes.Any() && info.GrowthTerrainTypes.Contains(self.World.Map.GetTerrainInfo(self.Location).Type))
```
Comment (RoosterDragon) [inline OpenRA.Mods.D2k/Traits/SpiceBloom.cs]: Checking for `Any` here is redundant - did you mean to check something else? If not you can remove it.

Comment (pchote) [inline OpenRA.Mods.D2k/Traits/SpiceBloom.cs]: I think this should be `if (any && !contains)`, right?
We want to exit early if the current terrain type isn't in the set, but only if the set isn't empty.

File: OpenRA.Mods.D2k/Traits/SpiceBloom.cs (hunk 5 above)
Comment (atlimit8) [inline OpenRA.Mods.D2k/Traits/SpiceBloom.cs]: `HashSet<string>`

File: OpenRA.Mods.D2k/Traits/SpiceBloom.cs [hunk 9]
```diff
@@ -91,45 +103,50 @@ public void Tick(Actor self)
 
 		public void Killed(Actor self, AttackInfo e)
 		{
-			var pieces = self.World.SharedRandom.Next(info.Pieces[0], info.Pieces[1]) * ticks / growTicks;
-			if (pieces < info.Pieces[0])
-				pieces = info.Pieces[0];
-
-			for (var i = 0; pieces > i; i++)
+			if (!string.IsNullOrEmpty(info.Weapon))
```
Comment (pchote) [inline OpenRA.Mods.D2k/Traits/SpiceBloom.cs]: `if (isnulllorempty) return;` would be cleaner.

File: OpenRA.Mods.D2k/Traits/SpiceBloom.cs [hunk 10]
```diff
@@ -78,6 +84,12 @@ public SpiceBloom(ActorInitializer init, SpiceBloomInfo info)
 
 		public void Tick(Actor self)
 		{
+			if (!self.World.Map.Contains(self.Location))
+				return;
+
+			if (info.GrowthTerrainTypes.Any() && !info.GrowthTerrainTypes.Contains(self.World.Map.GetTerrainInfo(self.Location).Type))
```
Comment (RoosterDragon) [inline OpenRA.Mods.D2k/Traits/SpiceBloom.cs]: Minor perf nit: Prefer `.Count > 0` since `Any` needs to allocate and this is called every tick.

File: OpenRA.Mods.D2k/Traits/SpiceBloom.cs [hunk 11]
```diff
@@ -91,45 +103,50 @@ public void Tick(Actor self)
...
-				var cell = cells.Take(info.Range).SkipWhile(p => resLayer.GetResource(p) == resType && resLayer.IsFull(p)).Cast<CPos?>().RandomOrDefault(self.World.SharedRandom);
-				if (cell == null)
-					cell = cells.Take(info.Range).Random(self.World.SharedRandom);
+				var pieces = self.World.SharedRandom.Next(info.Pieces[0], info.Pieces[1]) * ticks / growTicks;
+				if (pieces < info.Pieces[0])
+					pieces = info.Pieces[0];
 
-				var args = new ProjectileArgs
+				var cells = self.World.Map.FindTilesInAnnulus(self.Location, 1, info.Range)
+					.SkipWhile(p => resLayer.GetResource(p) == resType && resLayer.IsFull(p));
```
Comment (pchote) [inline OpenRA.Mods.D2k/Traits/SpiceBloom.cs]: This may need to be `ToList`ed in order to get the benefits of it.

File: OpenRA.Mods.D2k/Traits/SpiceBloom.cs (hunk 9 above)
Comment (reaperrr) [inline OpenRA.Mods.D2k/Traits/SpiceBloom.cs]: I think you're forgetting the spawning that starts at line 152 and is unrelated to the weapon.

File: OpenRA.Mods.D2k/Traits/SpiceBloom.cs [hunk 12]
```diff
@@ -91,45 +103,50 @@ public void Tick(Actor self)
...
+
+				for (var i = 0; pieces > i; i++)
```
Comment (pchote) [inline OpenRA.Mods.D2k/Traits/SpiceBloom.cs]: Style nit: `i < pieces`

File: OpenRA.Mods.D2k/Traits/SpiceBloom.cs (hunk 9 above)
Comment (pchote) [inline OpenRA.Mods.D2k/Traits/SpiceBloom.cs]: Ah, I missed that.  Still, this is a sign that it may want to be pulled out into a SeedResources method?

File: OpenRA.Mods.D2k/Traits/SpiceBloom.cs (hunk 11 above)
Comment (reaperrr) [inline OpenRA.Mods.D2k/Traits/SpiceBloom.cs]: Basically `.Where(p => resLayer.GetResource(p) != resType || !resLayer.IsFull(p)).ToList()`?

Comment (GraionDilach) [inline OpenRA.Mods.D2k/Traits/SpiceBloom.cs]: Also see #10062.

File: OpenRA.Mods.D2k/Traits/SpiceBloom.cs [hunk 13]
```diff
@@ -91,45 +103,50 @@ public void Tick(Actor self)
...
 				{
-					Weapon = self.World.Map.Rules.Weapons[info.Weapon.ToLowerInvariant()],
-					Facing = 0,
+					var cell = cells.Cast<CPos?>().RandomOrDefault(self.World.SharedRandom);
+					if (cell == null)
+						cell = cells.Random(self.World.SharedRandom);
```
Comment (RoosterDragon) [inline OpenRA.Mods.D2k/Traits/SpiceBloom.cs]: What are these three lines trying to do exactly? `RandomOfDefault` will return null only on an empty collection, so repeating it with `Random` will just throw which seems odd.

As written, you could simplify the current logic down to `cell = cells.Random(self.World.SharedRandom)` for the same effect.

File: OpenRA.Mods.D2k/Traits/SpiceBloom.cs (hunk 11 above)
Comment (RoosterDragon) [inline OpenRA.Mods.D2k/Traits/SpiceBloom.cs]: `self.World.Map.FindTilesInAnnulus` will only find cells in the world by default, so you don't need the same check as #10062.

Comment (GraionDilach) [inline OpenRA.Mods.D2k/Traits/SpiceBloom.cs]: Ah, sorry then.

File: OpenRA.Mods.D2k/Traits/SpiceBloom.cs (hunk 13 above)
Comment (reaperrr) [inline OpenRA.Mods.D2k/Traits/SpiceBloom.cs]: Tbh I didn't (and still don't) fully understand this piece of code from #8987, so I just avoided touching it too much.

Comment (RoosterDragon) [inline OpenRA.Mods.D2k/Traits/SpiceBloom.cs]: Looks like the old code was attempting to take a random cell in range but skipping over anything already full of resource. If there were no such cells (everything was already maxed) it just used any random cell in range instead.

Try moving the `SkipWhile` currently on the `cells` variable back onto the first random search line - that ought to make it work like the original code again as best I can tell.

**Comment by obrakmann** [comment]
Looks fine to me. :+1:

Can you please squash those last four commits into the earlier ones before merging? Thanks!

**Comment by reaperrr** [comment]
Squashed and reordered commits.

Note: There were some unavoidable yaml rebase conflicts while squashing. I'm quite sure i caught them all, but I still recommend giving this a last quick look-over before merging.

**Comment by obrakmann** [comment]
Nope, still looks all good, in-game too.

**Comment by obrakmann** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog/_compare/958c08cff8820a3c01d0c3d0d26cf7b1fa4ab83a)
